File: grpc-server/controller/todo.go
package controller

import (
	"context"
	api "github.com/eminoz/todo/pb"
)

type TodoService interface {
	CreateNewTodo(ctx context.Context, todo *api.Todo) (*api.ResTodo, error)
	AddNewTodo(ctx context.Context, todo *api.Todo) (*api.ResTodo, error)
	GetTodos(ctx context.Context, id *api.CustomerId) (*api.ResTodo, error)
	DeleteOneTodo(ctx context.Context, id *api.RemoveTodo) (*api.ResDeleteTodo, error)
	DeleteAllTodos(ctx context.Context, id *api.CustomerId) (*api.ResDeleteTodo, error)
}
type Todo struct {
	TodoService TodoService
}

func (t Todo) CreateTodo(ctx context.Context, todo *api.Todo) (*api.ResTodo, error) {
	newTodo, err := t.TodoService.CreateNewTodo(ctx, todo)
	if err != nil {
		panic(err)
	}

	return newTodo, nil
}

func (t Todo) AddNewToto(ctx context.Context, todo *api.Todo) (*api.ResTodo, error) {
	addNewTodo, err := t.TodoService.AddNewTodo(ctx, todo)
	if err != nil {
		panic(err)
	}
	return addNewTodo, err
}

func (t Todo) GetTodo(ctx context.Context, id *api.CustomerId) (*api.ResTodo, error) {
	todos, err := t.TodoService.GetTodos(ctx, id)
	if err != nil {
		panic(err)
	}
	return todos, err
}

func (t Todo) DeleteTodo(ctx context.Context, id *api.RemoveTodo) (*api.ResDeleteTodo, error) {
	deleteOneTodo, err := t.TodoService.DeleteOneTodo(ctx, id)
	if err != nil {
		panic(err)
	}
	return deleteOneTodo, err
}

func (t Todo) DeleteAllTodo(ctx context.Context, id *api.CustomerId) (*api.ResDeleteTodo, error) {
	todos, err := t.TodoService.DeleteAllTodos(ctx, id)
	if err != nil {
		panic(err)
	}
	return todos, err
}

File: grpc-client/api/todo.js
const grpc = require("@grpc/grpc-js");
const protoLoader = require("@grpc/proto-loader");
var PROTO_PATH = __dirname + "../../../todo.proto";
const packageDefinition = protoLoader.loadSync(PROTO_PATH, {});
const grpcObject = grpc.loadPackageDefinition(packageDefinition).userproto;

const todoService = new grpcObject.TodoService(
  "localhost:4040",
  grpc.credentials.createInsecure()
);

module.exports = (app) => {
  app.post("/createnewtodo/:id", (req, res, next) => {
    const CustomerId = req.params.id;
    const { Title } = req.body;

    todoService.CreateTodo({ Title, CustomerId }, (err, response) => {
      const data = {
        todos: response.Title,
        CustomerId: response.CustomerId,
      };
      res.json({ data });
    });
  });
  app.post("/addnewtodo/:id", (req, res, next) => {
    const CustomerId = req.params.id;
    const { Title } = req.body;
    todoService.AddNewToto({ Title, CustomerId }, (err, response) => {
      const data = {
        todos: response.Title,
        CustomerId: response.CustomerId,
      };
      res.json({ data });
    });
  });
  app.get("/gettodosbycustomerid/:id", (req, res, next) => {
    const CustomerId = req.params.id;
    todoService.GetTodo({ Id: CustomerId }, (err, response) => {
      console.log(response);
      const data = {
        todos: response.Title,
        CustomerId: response.CustomerId,
      };
      res.json({ data });
    });
  });
  app.post("/deletetodo/:id", (req, res, next) => {
    const CustomerId = req.params.id;
    const { Title } = req.body;
    todoService.DeleteTodo({ Title, CustomerId }, (err, response) => {
      const data = {
        title: response.Title,
        message: response.Message,
      };
      res.json({ data });
    });
  });
  app.get("/deletealltodos/:id", (req, res, next) => {
    const CustomerId = req.params.id;
    todoService.DeleteAllTodo({ Id: CustomerId }, (err, response) => {
      const data = {
        title: response.Title,
        message: response.Message,
      };
      res.json({
        data,
      });
    });
  });
};
